Adding more details about downloaded geoip database files

masseyke opened the pull request:
This add new fields to the `_ingest/geoip/stats` response so that it looks something like:
```
{
    "stats": {
        "successful_downloads": 3,
        "failed_downloads": 0,
        "total_download_time": 6110,
        "databases_count": 3,
        "skipped_updates": 0,
        "expired_databases": 0,
        "downloader_attempts": [
            {
                "name": "GeoLite2-ASN.mmdb",
                "last_success": {
                    "archive_md5": "a10c9cc703c58a1bd08c5663217fe8ef",
                    "download_date": "2024-06-07T20:28:26.371Z",
                    "download_date_in_millis": 1717792106371,
                    "download_time_in_millis": 2237,
                    "download_time": "2.2s",
                    "source": "geoip.elastic.co",
                    "build_date": "2024-06-04T13:35:04.000Z",
                    "build_date_in_millis": 1717508104000
                }
            },
            {
                "name": "GeoLite2-City.mmdb",
                "last_success": {
                    "archive_md5": "b64c6fc17beb6d2cbcee98ec0fd795da",
                    "download_date": "2024-06-07T20:28:30.398Z",
                    "download_date_in_millis": 1717792110398,
                    "download_time_in_millis": 2648,
                    "download_time": "2.6s",
                    "source": "geoip.elastic.co",
                    "build_date": "2024-06-04T13:50:26.000Z",
                    "build_date_in_millis": 1717509026000
                }
            },
            {
                "name": "GeoLite2-Country.mmdb",
                "last_success": {
                    "archive_md5": "19835a83eed6886737da18694e870954",
                    "download_date": "2024-06-07T20:28:35.191Z",
                    "download_date_in_millis": 1717792115191,
                    "download_time_in_millis": 1225,
                    "download_time": "1.2s",
                    "source": "geoip.elastic.co",
                    "build_date": "2024-06-04T13:54:13.000Z",
                    "build_date_in_millis": 1717509253000
                }
            }
        ]
    },
    "nodes": {
        "DkkEL3chQcauySLjzAemoA": {
            "databases": [
                {
                    "name": "GeoIP2-City-Test.mmdb",
                    "source": "config",
                    "archive_md5": null,
                    "md5": "f858c72974318926cdbcd256cea3417d",
                    "build_date": "2020-12-29T17:44:40.000Z",
                    "build_date_in_millis": 1609263880000,
                    "type": "GeoIP2-City"
                },
                {
                    "name": "GeoLite2-ASN.mmdb",
                    "source": "downloader",
                    "archive_md5": "a10c9cc703c58a1bd08c5663217fe8ef",
                    "md5": "2bb15706a46e2ebf82ae3bb3b86fda58",
                    "build_date": "2024-06-04T12:03:26.000Z",
                    "build_date_in_millis": 1717502606000,
                    "type": "GeoLite2-ASN"
                },
                {
                    "name": "GeoLite2-City.mmdb",
                    "source": "downloader",
                    "archive_md5": "b64c6fc17beb6d2cbcee98ec0fd795da",
                    "md5": "0beaaac3bc36a2fae77d1158216da8ed",
                    "build_date": "2024-06-04T13:35:07.000Z",
                    "build_date_in_millis": 1717508107000,
                    "type": "GeoLite2-City"
                },
                {
                    "name": "GeoLite2-Country.mmdb",
                    "source": "downloader",
                    "archive_md5": "19835a83eed6886737da18694e870954",
                    "md5": "79e7e522dbe986ca913464f2894f180e",
                    "build_date": "2024-06-04T13:46:29.000Z",
                    "build_date_in_millis": 1717508789000,
                    "type": "GeoLite2-Country"
                }
            ],
            "files_in_temp": [
                "GeoLite2-City.mmdb",
                "GeoLite2-Country.mmdb_COPYRIGHT.txt",
                "GeoLite2-City.mmdb_README.txt",
                "GeoLite2-ASN.mmdb_COPYRIGHT.txt",
                "GeoLite2-City.mmdb_elastic-geoip-database-service-agreement-LICENSE.txt",
                "GeoLite2-ASN.mmdb_elastic-geoip-database-service-agreement-LICENSE.txt",
                "GeoLite2-Country.mmdb_elastic-geoip-database-service-agreement-LICENSE.txt",
                "GeoLite2-City.mmdb_LICENSE.txt",
                "GeoLite2-ASN.mmdb_LICENSE.txt",
                "GeoLite2-ASN.mmdb",
                "GeoLite2-Country.mmdb_LICENSE.txt",
                "GeoLite2-City.mmdb_COPYRIGHT.txt",
                "GeoLite2-Country.mmdb"
            ],
            "config_databases": [
                "GeoLite2-City.mmdb",
                "GeoIP2-City-Test.mmdb"
            ],
            "cache_stats": {
                "count": 0,
                "hits": 0,
                "misses": 0,
                "evictions": 0,
                "hits_time": "0s",
                "hits_time_in_millis": 0,
                "misses_time": "0s",
                "misses_time_in_millis": 0
            }
        }
    }
}
```
The entire `downloader_attempts` section is new, reporting on the attempts to download databases from a remote system. The `databases` section now has the new fields `source`, `md5_archive`, `md5`, `build_date_in_millis`, `build_date` (if the `human` request parameter is given), and `type`.
Note that the configured database `GeoLite2-City.mmdb` does not appear in the `databases` section -- this is because downloaded databases take priority over configured ones, and there is a downloaded `GeoLite2-City.mmdb`. The `databases` section only shows databases on disk that will actually be used.

masseyke committed:
Adding more details about downloaded geoip files (work in progress)

masseyke committed:
unit tests now pass, removed unneeded fields

elasticsearchmachine commented:
Hi @masseyke, I've created a changelog YAML for you.

masseyke committed:
Update docs/changelog/108985.yaml

masseyke committed:
cleaning up

masseyke committed:
Merge branch 'geoip-download-stats-more-details' of github.com:masseyke/elasticsearch into geoip-download-stats-more-details

masseyke committed:
adding serialization unit tests

masseyke committed:
adding documentation

masseyke committed:
minor fixes

masseyke committed:
merging main

masseyke committed:
fixing GeoIpDownloaderCliIT

masseyke committed:
improving GeoIpDownloaderStatsIT

masseyke committed:
adding to the yaml rest test

joegallo committed:
Merge branch 'main' into geoip-download-stats-more-details

joegallo commented:
```
    "nodes": {
        "6mkRmvVpR8yVzaAxa7Gc3w": {
            "databases": [
                {
                    "name": "GeoLite2-ASN.mmdb",
                    "md5": "bb4934270944faf4928e2963d8e99cda",
                    "build_date": "2024-05-21T23:05:55.000Z",
                    "build_date_in_millis": 1716332755000,
                    "type": "GeoLite2-ASN"
                },
```

^ does a section like that only show up for actually downloaded databases? I think I'm running the code from this branch, and I copied a file into `config/ingest-geoip`manually (and I disabled the downloader). Here's what I'm seeing:

```
GET _ingest/geoip/stats?human
{
  "stats" : {
    "successful_downloads" : 0,
    "failed_downloads" : 0,
    "total_download_time" : 0,
    "databases_count" : 0,
    "skipped_updates" : 0,
    "expired_databases" : 0,
    "downloader_attempts" : [ ]
  },
  "nodes" : {
    "XrYC3BGZQjumVHtz6-4vGA" : {
      "config_databases" : [
        "GeoLite2-ASN.mmdb"
      ],
      "cache_stats" : {
        "count" : 0,
        "hits" : 0,
        "misses" : 0,
        "evictions" : 0,
        "hits_time" : "0s",
        "hits_time_in_millis" : 0,
        "misses_time" : "0s",
        "misses_time_in_millis" : 0
      }
    }
  }
}
```

joegallo commented:
I really like all of this! There's some little details to iron out, but I'm very happy with this new capability in the large.

masseyke commented:
> What actually sets the `source` in the end? The description on the PR has `"source": "storage.googleapis.com"` (which I wasn't absolutely thrilled with) but when I ran thinks locally I saw [geoip.elastic.co]
> And that's actually more in line with what I'd want, so I just want to understand the planned behavior better. I think everything is probably fine, but I want to understand.

Yeah `geoip.elastic.co` is the intended behavior -- the place the data is actually coming from. I only put `storage.googleapis.com` (the location of the manifest, not the data) in the description because I wasn't paying close attention. I'll fix that now (the description).

masseyke committed:
adding md5 for the mmdb files

masseyke committed:
merging main

masseyke committed:
merging main

masseyke committed:
clarifying that downloader databases take precedence over config databases

masseyke committed:
merging main

masseyke committed:
merging main

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

masseyke committed:
only return one database per name

masseyke committed:
asserting that the correct database is returned

masseyke committed:
fixing a GeoIpStatsAction bug  caused by deleting config databases

masseyke committed:
merging main

masseyke committed:
fixing the log level in an error message

masseyke committed:
improving an error message

masseyke committed:
spotless

masseyke committed:
merging main

masseyke committed:
Merge branch 'main' into geoip-download-stats-more-details

masseyke committed:
Merge branch 'main' into geoip-download-stats-more-details

joegallo committed:
A slight simplification

joegallo committed:
Fix a typo

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/stats/GeoIpDownloaderStats.java:
```diff
@@ -44,6 +53,15 @@ public GeoIpDownloaderStats(StreamInput in) throws IOException {
+                downloadedDatabaseInfos.put(in.readString(), new DownloadedDatabaseInfo(in));
+            }
+        } else {
+            downloadedDatabaseInfos = null;
```
Would it tidy the logic elsewhere very much if this were an empty map instead of null?

masseyke replied:
I think that the reason I did that is to avoid sending "downloader_attempts" in the xcontent response to nodes that don't know anything about it. But nothing actually parses this class's xcontent other than unit tests, so maybe that's unnecessary. I'll remove this and make sure nothing breaks.

joegallo committed:
Rectification of names

A database is 'retrieved' when it is reconstructed as a local file on
some node in the cluster using the chunks from the geoip_databases
index. Since this class is also used to report on the info for
'config/' databases that were not ever retrieved, then let's not have
'Retrieved' in the class name.

joegallo commented on modules/ingest-geoip/src/test/java/org/elasticsearch/ingest/geoip/stats/GeoIpDownloaderStatsSerializingTests.java:
```diff
@@ -11,29 +11,87 @@
+            (int) a[3],
+            (int) a[4],
+            a[5] == null ? 0 : (int) a[5],
+            a[6] == null ? null : (List<DownloadedDatabaseInfo>) a[6]
```
My IntelliJ doesn't like this `null` here -- if this were to run, then I think we'd NPE in the `GeoIpDownloaderStats` constructors.

joegallo commented on modules/ingest-geoip/src/main/java/org/elasticsearch/ingest/geoip/DatabaseNodeService.java:
```diff
@@ -492,8 +499,60 @@ void retrieveDatabase(
+                )
+            );
+        }
+        return new HashSet<>(allDatabases.values());
```
A minor nit, but what do you think about `return Set.copyOf(allDatabases.values())` or some-such here instead?

masseyke committed:
getting rid of unreachable code

masseyke committed:
Merge branch 'geoip-download-stats-more-details' of github.com:masseyke/elasticsearch into geoip-download-stats-more-details

masseyke committed:
removing support for null downloadedDatabaseInfos

masseyke committed:
using Set.copyOf rather than new HashSet

joegallo committed:
Merge branch 'main' into geoip-download-stats-more-details

masseyke committed:
Merge branch 'main' into geoip-download-stats-more-details

masseyke committed:
merging main

masseyke commented:
> I'm not sure whether `downloader_attempts` is the best name for this

How about `download_attempts`? Or `database_download_attempts`?
